### DISCUSSION TITLE
Fast Refresh not working with page that is wrapped in a HOC

### DISCUSSION THREAD

**Discussion by PhanDungTri** [vercel/next.js#37405] [Help] [answered]
Hello, I want to secure some pages in my website so I wrap them into a HOC that will check if user has logged in.

![image](https://user-images.githubusercontent.com/26342804/171654540-2b2c068a-81fc-435f-bf57-5d013a895492.png)

Here is my HOC implementation.

![image](https://user-images.githubusercontent.com/26342804/171655236-9c9c6750-2b0e-408d-baed-ae1e355910b8.png)

But when I make some changes in those pages, Next warns me about Fast Refresh will perform a full reload.

![image](https://user-images.githubusercontent.com/26342804/171654817-71c54537-56ed-4810-b6e1-d7e5aec0ced3.png)

How can I fix this problem?

**Comment by icyJoseph** [comment, marked as answer]
### Solution

It is good to avoid anonymous functions because otherwise, the algorithm trying to figure out what to preserve can't do that very thing, specifically when HoCs are involved, or at least that's my take on it.

In your case this snippet should work. That is, give a name to the Component definition (returned function) from `withAuthRequirement`. In this case I call mine, `Enhanced`. I'd also recommend reading about [naming conventions for HoCs](https://reactjs.org/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging).

```tsx
const withNothing = <PageProps,>(Page: NextPage<PageProps>) => {
  return function Enhanced(props: PageProps) {
    const factor = 3;
    const mult = 2 * factor;

    return <Page {...props} />;
  };
};
```

or

```tsx
const withNothing = <PageProps,>(Page: NextPage<PageProps>) => {
  function Enhanced(props: PageProps) {
    const factor = 3;
    const mult = 2 * factor;

    return <Page {...props} />;
  }

  Enhanced.displayName = `withNothing(${
    Page.displayName || Page.name || "NextPage"
  })`;

  return Enhanced;
};
```

### Extras

I'd also suggest to avoid:

```tsx
const Home = withNothing(() => {
  return (
    <div>
      Home <Link href="/blog">blog</Link>
    </div>
  );
});

export default Home;
```

and do this instead, for cleanliness, and to keep things decoupled.

```tsx
const Home = () => {
  return (
    <div>
      Home <Link href="/blog">blog</Link>
    </div>
  );
};

export default withNothing(Home)
```
